src/Form/FormAnalyzer.js

```diff
@@ -229,6 +232,92 @@ class FormAnalyzer {
+     * Takes an element and returns all its children that are text-only nodes
+     * @param {HTMLElement|Element} element
+     * @param {number} maxDepth
+     * @param {number} currentDepth
+     * @returns {HTMLElement[]|Element[]}
+     */
+    getElementsWithOnlyTextChild(element, maxDepth = 2, currentDepth = 0) {
+        // Array to collect elements with only text child nodes
+        const elementsWithTextChild = [];
+
+        // If we've reached the max depth, stop traversing further
+        if (currentDepth > maxDepth) {
+            return elementsWithTextChild;
+        }
+
+        // Check if the current element has only one text child node
+        if (element.nodeType === Node.ELEMENT_NODE) {
+            const childNodes = element.childNodes;
+
+            if (childNodes.length === 1 && childNodes[0].nodeType === Node.TEXT_NODE) {
+                elementsWithTextChild.push(element);
+            }
+        }
+
+        // Recurse through each child element and collect matching elements
+        for (const child of element.children) {
+            // Recursively get elements from child elements, increasing depth by 1
+            elementsWithTextChild.push(...this.getElementsWithOnlyTextChild(child, maxDepth, currentDepth + 1));
+        }
+
+        return elementsWithTextChild;
+    }
+
+    evaluateFormHeaderSignals() {
+        const isVisuallyBeforeForm = (el) => el.getBoundingClientRect().top < this.form.getBoundingClientRect().top;
+
+        const isHeaderSized = (el) => {
+            if (el instanceof HTMLHeadingElement) {
+                return true;
+            }
+
+            const computedStyle = window.getComputedStyle(el);
+            const fontWeight = computedStyle.fontWeight;
+            const isRelativelyTall = parseFloat(computedStyle.height) / this.form.clientHeight > 0.1;
+            if (isRelativelyTall && (fontWeight === 'bold' || parseFloat(fontWeight) >= 700)) {
+                return true;
+            }
+        };
+
+        const allSiblings = Array.from(this.form.parentElement?.children ?? [])
+            .filter((element) => element !== this.form)
+            .map((element) => this.getElementsWithOnlyTextChild(element))
+            .flat();
+
+        if (allSiblings.length === 0) return false;
+
+        allSiblings.forEach((element) => {
+            if (element instanceof HTMLElement && isVisuallyBeforeForm(element) && isHeaderSized(element)) {
+                const string = element.textContent?.trim();
+                if (string) {
+                    if (safeRegexTest(/^(sign[- ]?in|log[- ]?in)$/i, string)) {
+                        return this.decreaseSignalBy(3, 'Strong login signal above form');
+                    } else if (safeRegexTest(/^(sign[- ]?up)$/i, string)) {
+                        return this.increaseSignalBy(3, 'Strong signup signal above form');
+                    }
+                }
+            }
+        });
+    }
+
+    evaluatePasswordHints() {
```

Comment:
Looping through divs and spans is complex and potentially expensive. What if instead we looked at the whole form `textContent` and run the regex once per form? We'd need to increase the `TEXT_LENGTH_CUTOFF`, or have a carveout for this situation, like making `safeRegexTest` accept a parameter and then have another higher cutoff constant we can use here.